From a9888e2dca89f6fa8b01287e4503a68f29f65e92 Mon Sep 17 00:00:00 2001
From: Vasiliy Stelmachenok <cabopust@yandex.ru>
Date: Sun, 3 Oct 2021 11:41:14 +0300
Subject: [PATCH] NVreg Improvements v2

This patch edits the default values of the NVIDIA kernel module
parameters, and makes the following changes:

NVreg_UsePageAttributeTable=1 (Default 0) - Activating the better
memory management method (PAT). The PAT method creates a partition type table
at a specific address mapped inside the register and utilizes the memory architecture
and instruction set more efficiently and faster.
If your system can support this feature, it should improve CPU performance.

NVreg_EnablePCIeGen3=1 (Default 0) - Enable PCIe Gen 3.x support.
If your graphics card supports PCIe 3.x, this driver module option
can active the high-speed 8 GT bus.
By default the Nvidia driver is set to use PCIe Gen 2.x for compatibility reasons.

NVreg_InitializeSystemMemoryAllocations=0 (Default 1) - Disables
clearing system memory allocation before using it for the GPU.
Potentially improves performance, but at the cost of increased security risks.
Write "options nvidia NVreg_InitializeSystemMemoryAllocations=1" in /etc/modprobe.d/nvidia.conf,
if you want to return the default value.
Note: It is possible to use more video memory (?)

NVreg_EnableStreamMemOPs=1 (Default 0) - Activates the support for
CUDA Stream Memory Operations in user-mode applications.

NVreg_PreserveVideoMemoryAllocations=1 (Default 0) - Prompts the NVIDIA kernel module to save
and restore all video memory allocations during all system power management cycles,
i.e. suspend/resume and hibernate/resume.
This helps prevent applications from freezing or rendering errors after resume.

NVreg_DynamicPowerManagement=Ox02 (Default 0x03) - Enables dynamic power management
where the NVIDIA GPU can completely shut down when it is not assigned tasks,
resulting in better power savings.
This feature works only on laptops with Turing or higher, Intel Coffeelake processors, and Linux 4.18+ kernel.
In other cases this option has no effect.
You can check if your hardware supports this feature with the following command:

$ cat /proc/driver/nvidia/gpus/0000:01:00.0/power

Read more about it: https://download.nvidia.com/XFree86/Linux-x86_64/470.74/README/dynamicpowermanagement.html

NVreg_EnableS0ixPowerManagement=1 (Default 0) - Enables putting the GPU video memory into self-updating mode
during s2idle system suspend. This feature is only supported on platforms with S0ix-based power management,
which can be verified via the following command:

$ grep 'Video Memory Self Refresh' /proc/driver/nvidia/gpus/0000\:01\:00.0/power


If you want to learn more about the NVIDIA driver module parameters,
you can go to the Gentoo Wiki page or view the source code of the nv-reg.h file.

---
 kernel/nvidia/nv-reg.h | 14 +++++++-------
 1 file changed, 7 insertions(+), 7 deletions(-)

diff --git a/kernel/nvidia/nv-reg.h b/kernel/nvidia/nv-reg.h
index 1ee7473..4482519 100644
--- a/kernel/nvidia/nv-reg.h
+++ b/kernel/nvidia/nv-reg.h
@@ -774,20 +774,20 @@ NV_DEFINE_REG_ENTRY_GLOBAL(__NV_MODIFY_DEVICE_FILES, 1);
 NV_DEFINE_REG_ENTRY(__NV_DEVICE_FILE_UID, 0);
 NV_DEFINE_REG_ENTRY(__NV_DEVICE_FILE_GID, 0);
 NV_DEFINE_REG_ENTRY(__NV_DEVICE_FILE_MODE, 0666);
-NV_DEFINE_REG_ENTRY(__NV_INITIALIZE_SYSTEM_MEMORY_ALLOCATIONS, 1);
-NV_DEFINE_REG_ENTRY(__NV_USE_PAGE_ATTRIBUTE_TABLE, ~0);
+NV_DEFINE_REG_ENTRY(__NV_INITIALIZE_SYSTEM_MEMORY_ALLOCATIONS, 0);
+NV_DEFINE_REG_ENTRY(__NV_USE_PAGE_ATTRIBUTE_TABLE, 1);
 NV_DEFINE_REG_ENTRY(__NV_REGISTER_FOR_ACPI_EVENTS, 1);
-NV_DEFINE_REG_ENTRY(__NV_ENABLE_PCIE_GEN3, 0);
+NV_DEFINE_REG_ENTRY(__NV_ENABLE_PCIE_GEN3, 1);
 NV_DEFINE_REG_ENTRY(__NV_ENABLE_MSI, 1);
 NV_DEFINE_REG_ENTRY(__NV_TCE_BYPASS_MODE, NV_TCE_BYPASS_MODE_DEFAULT);
-NV_DEFINE_REG_ENTRY(__NV_ENABLE_STREAM_MEMOPS, 0);
+NV_DEFINE_REG_ENTRY(__NV_ENABLE_STREAM_MEMOPS, 1);
 NV_DEFINE_REG_ENTRY(__NV_RM_PROFILING_ADMIN_ONLY_PARAMETER, 1);
-NV_DEFINE_REG_ENTRY(__NV_PRESERVE_VIDEO_MEMORY_ALLOCATIONS, 0);
+NV_DEFINE_REG_ENTRY(__NV_PRESERVE_VIDEO_MEMORY_ALLOCATIONS, 1);
 
-NV_DEFINE_REG_ENTRY(__NV_ENABLE_S0IX_POWER_MANAGEMENT, 0);
+NV_DEFINE_REG_ENTRY(__NV_ENABLE_S0IX_POWER_MANAGEMENT, 1);
 NV_DEFINE_REG_ENTRY(__NV_S0IX_POWER_MANAGEMENT_VIDEO_MEMORY_THRESHOLD, 256);
 
-NV_DEFINE_REG_ENTRY(__NV_DYNAMIC_POWER_MANAGEMENT, 3);
+NV_DEFINE_REG_ENTRY(__NV_DYNAMIC_POWER_MANAGEMENT, 2);
 NV_DEFINE_REG_ENTRY(__NV_DYNAMIC_POWER_MANAGEMENT_VIDEO_MEMORY_THRESHOLD, 200);
 NV_DEFINE_REG_ENTRY(__NV_ENABLE_GPU_FIRMWARE, NV_REG_ENABLE_GPU_FIRMWARE_DEFAULT_VALUE);
 
-- 
2.33.0

